server/src/main/java/io/unitycatalog/server/service/PermissionService.java

```diff
@@ -0,0 +1,278 @@
+  }
+
+  @Patch("/schema/{name}")
+  @AuthorizeExpression(
+      "#authorize(#principal, #metastore, OWNER) || #authorize(#principal, #schema, OWNER)")
+  @AuthorizeKey(METASTORE)
+  public HttpResponse updateSchemaAuthorization(
+      @Param("name") @AuthorizeKey(SCHEMA) String name, UpdatePermissions request) {
+    return updateAuthorization(SCHEMA, name, request);
+  }
+
+  @Patch("/table/{name}")
+  @AuthorizeExpression(
+      "#authorize(#principal, #metastore, OWNER) || #authorize(#principal, #table, OWNER)")
+  @AuthorizeKey(METASTORE)
+  public HttpResponse updateTableAuthorization(
+      @Param("name") @AuthorizeKey(TABLE) String name, UpdatePermissions request) {
+    return updateAuthorization(TABLE, name, request);
+  }
+
+  @Patch("/function/{name}")
+  @AuthorizeExpression(
+      "#authorize(#principal, #metastore, OWNER) || #authorize(#principal, #function, OWNER)")
+  @AuthorizeKey(METASTORE)
+  public HttpResponse updateFunctionAuthorization(
+      @Param("name") @AuthorizeKey(FUNCTION) String name, UpdatePermissions request) {
+    return updateAuthorization(FUNCTION, name, request);
+  }
+
+  @Patch("/volume/{name}")
+  @AuthorizeExpression(
+      "#authorize(#principal, #metastore, OWNER) || #authorize(#principal, #volume, OWNER)")
+  @AuthorizeKey(METASTORE)
+  public HttpResponse updateVolumeAuthorization(
+      @Param("name") @AuthorizeKey(VOLUME) String name, UpdatePermissions request) {
+    return updateAuthorization(VOLUME, name, request);
+  }
+
+  @Patch("/registered-model/{name}")
+  @AuthorizeExpression(
+      "#authorize(#principal, #metastore, OWNER) || #authorize(#principal, #registered_model, OWNER)")
+  @AuthorizeKey(METASTORE)
+  public HttpResponse updateRegisteredModelAuthorization(
+      @Param("name") @AuthorizeKey(REGISTERED_MODEL) String name, UpdatePermissions request) {
+    return updateAuthorization(REGISTERED_MODEL, name, request);
+  }
+
+  private HttpResponse updateAuthorization(
+      SecurableType securableType, String name, UpdatePermissions request) {
+    UUID resourceId = getResourceId(securableType, name);
+    List<PermissionsChange> changes = request.getChanges();
+    Set<UUID> principalIds = new HashSet<>();
+    changes.forEach(
+        change -> {
+          String principal = change.getPrincipal();
+          User user = USER_REPOSITORY.getUserByEmail(principal);
+          UUID principalId = UUID.fromString(Objects.requireNonNull(user.getId()));
+          principalIds.add(principalId);
+          change
+              .getAdd()
+              .forEach(
+                  privilege ->
+                      //  Privileges should always be a superset of Privilege so this _should_
+                      // always be non-null but let's be safe anyway.
+                      Optional.ofNullable(Privileges.fromPrivilege(privilege))
+                          .map(p -> authorizer.grantAuthorization(principalId, resourceId, p)));
```

Comment:
The grants and revokes are done one at a time in separate transactions by the authorizer, so the permission update isn't guaranteed to be atomic. If there's an error while doing a grant or revoke, the API returns an error, and the burden is on the user to clean up any partially applied grants. 

Is there a way to avoid this and make the application of multiple grants and revokes in a single API call atomic?

Reply:
The issue of lack of atomicity also surfaces when grants are applied as part of resource creation (and maybe later during update) -- the OWNER privilege is added after the resource create and if there's a failure in between, it can result in a created object that can't be deleted by the creator.

Reply:
Can you add a GH issue to track this?

Reply:
Can you add a GH issue to track this?